nx/lib/nx/lin_alg.ex

```diff
@@ -967,4 +967,59 @@ defmodule Nx.LinAlg do
       opts
     )
   end
+
+  @doc """
+  Produces the tensor taken to the given power by dot-product.
+
+  The input is always a square tensor and a non-negative integer,
+  and the output is a square tensor of the same dimensions as the input tensor.
+
+  ## Examples
+
+      iex> Nx.LinAlg.matrix_power(Nx.tensor([[1, 2], [3, 4]]), 0)
+      #Nx.Tensor<
+        s64[2][2]
+        [
+          [1, 0],
+          [0, 1]
+        ]
+      >
+
+      iex> Nx.LinAlg.matrix_power(Nx.tensor([[1, 2], [3, 4]]), 6)
+      #Nx.Tensor<
+        s64[2][2]
+        [
+          [5743, 8370],
+          [12555, 18298]
+        ]
+      >
+
+      iex> Nx.LinAlg.matrix_power(Nx.eye(3), 65535)
+      #Nx.Tensor<
+        s64[3][3]
+        [
+          [1, 0, 0],
+          [0, 1, 0],
+          [0, 0, 1]
+        ]
+      >
+
+  """
+  @doc from_backend: false
+  def matrix_power(matrix, power) when is_integer(power) and power < 0 do
+    matrix_power(Nx.LinAlg.invert(matrix), -power)
+  end
+  def matrix_power(matrix, power) when is_integer(power) do
+    {result, _exp_matrix} = power
+    |> Integer.digits(2)
+    |> Enum.reverse()
```

Comment:
If you extract this list to a variable, you can get it's length, and then iterate with the `Enum.reduce` by also accumulating the current index. Then you can skip the last superfluous dot-product by just returning the result matrix without calculating the next `exp_matrix`. This will also simplify things because you don't need to return a tuple in the last iteration, which enables you to remove the pattern match on `{result, _exp_matrix}`